src/karaoke.js

```diff
@@ -0,0 +1,101 @@
+    for (const speaker in speakerElements) {
+        speakerElements[speaker].quote.textContent = ''
+    }
+}
+
+
+/**
+ * Move the spoken word into the visible portion of the container. If the word is already fully visible, no change will
+ * be made.
+ */
+function moveWordIntoView(container, quote, wordSpan) {
+    const containerBoundingRect = container.getBoundingClientRect()
+    const wordSpanBoundingRect = wordSpan.getBoundingClientRect()
+    const bottomPadding = parseInt($(quote).css('padding-bottom'))
+    const topPadding = parseInt($(quote).css('padding-top'))
+    // Check if the word is outside of the quote container (even partially).
+    if ((
+        containerBoundingRect.top + topPadding > wordSpanBoundingRect.top ||
+        containerBoundingRect.bottom < wordSpanBoundingRect.bottom + bottomPadding
+    )) {
+        const quoteBoundingRect = quote.getBoundingClientRect()
+        // Shift the quote div up by the amount it would take to move the wordSpan up to the top of the container.
+        quote.style.top = `${quoteBoundingRect.top - wordSpanBoundingRect.top + topPadding}px`
+    }
+}
+
+
+/**
+ * Invoke this function from the `animations.js` script. `CURRENT_ANIMATION_INFO` is set as a global in that script and
+ * will be available to reference by the time this function is invoked.
+ */
+// eslint-disable-next-line no-unused-vars
+function updateKaraoke(karaokeAnimationInfo, animationTime) {
+    if (!CURRENT_ANIMATION_INFO.karaokeState) {
+        CURRENT_ANIMATION_INFO.karaokeState = {
+            currentQuoteStart: null,
+            stagedWords: []
+        }
+    }
+
+    const currentTime = animationTime + karaokeAnimationInfo.utterancesStartOffset
+    const currentQuote = getCurrentQuote(karaokeAnimationInfo.utterances, currentTime)
+
+    // There will be gaps between quotes. Don't do anything if there is not a quote at the `currentTime`.
+    if (currentQuote) {
+        const { container, quote } = karaokeAnimationInfo.speakerElements[currentQuote.speaker]
+
+        // Check if a new quote is starting. Clean up the previous quotes and set up the new quote.
+        if (currentQuote.start !== CURRENT_ANIMATION_INFO.karaokeState.currentQuoteStart) {
+            clearQuotes(karaokeAnimationInfo.speakerElements)
+            CURRENT_ANIMATION_INFO.karaokeState.currentQuoteStart = currentQuote.start
+            CURRENT_ANIMATION_INFO.karaokeState.stagedWords = []
+            currentQuote.words.forEach(word => {
+                const wordSpan = appendWord(quote, word, karaokeAnimationInfo.textVariant)
+                CURRENT_ANIMATION_INFO.karaokeState.stagedWords.push({element: wordSpan, ...word})
+            })
+        }
+
+        // Unstage all words that have been or are currently being spoken.
+        while (
+            CURRENT_ANIMATION_INFO.karaokeState.stagedWords.length > 0 &&
+            currentTime >= CURRENT_ANIMATION_INFO.karaokeState.stagedWords[0].start
```

Comment:
In which case does this condition come in to play? Should it always be true?

Reply:
@martingouy Good question! 

Once the word has been spoken and is made opaque, it's removed from the `stagedWords` array. Since this loop happens every animation frame, I was trying to limit the number of iterations it needs to go through. Ideally the contents of the loop will never need to execute more than once per animation frame, but I used a while loop in case there's lag in the animation where two words need to be made opaque at once it can get caught up in a single frame.

https://github.com/heyartifact/webflow/blob/05fd58aab191be004cf02ef1020f20d64fd8d49c/src/karaoke.js#L96-L97